### ISSUE TITLE
inconsistent results when using vw daemon

### ISSUE THREAD

**Issue by mustaphabenm** [VowpalWabbit/vowpal_wabbit#4067]
### Describe the bug

I'm trying to get infer on a vw model trained with cats_pdf using a daemon. the problem is that I'm not getting the same result for a specific context : you can see below that out of 100 trials I got 2 different results

```
1-28.591745:0.006451613,28.591745-29.11138:1.5459961,29.11138-32:0.006451613
  
1-29.07612:0.006451613,29.07612-29.595755:1.5459961,29.595755-32:0.006451613
```

### How to reproduce

1 - train a vw model  :

`vw --cats_pdf 64 --min_value 1 --max_value 32 -b 20 --chain_hash --nn 120 --dropout --progress 10000 --learning_rate 0.00003 --data dummy_data.vw --final_regressor model_new_params.pkl --bandwidth 0.25981749230238499`

2- create a daemon :

`vw --daemon --initial_regressor model_new_params.pkl  --epsilon 0.2 --chain_hash --quiet --testonly --port 8888 --num_children 10
`
3- communicate with the daemon : `python ping_daemon.py`

### Version

9.1.0 (git commit: bd4998c9e)

### OS

Linux

### Language

CLI

**Comment by martininstadeep** [comment]
We found that the reason for inconsistency comes from using dropout. Without dropout the predictions are consistent.

However, shouldn't there be an option to turn off dropout during inference?

**Comment by pmineiro** [comment]
> However, shouldn't there be an option to turn off dropout during inference?

Yes I'm surprised this doesn't happen already, and would consider it a bug.

For --lrq the dropout [activating --testonly produces mean field results](
https://github.com/VowpalWabbit/vowpal_wabbit/blob/eba08c42de6a1da8640bbfd2d75b6ff26bf4f18e/vowpalwabbit/core/src/reductions/lrq.cc#L87). However for --nn it [--testonly does nothing](https://github.com/VowpalWabbit/vowpal_wabbit/blob/c9cb804a74c94df4a00756c78f6e204526fcb7a9/vowpalwabbit/core/src/reductions/nn.cc#L164).

**Comment by JohnLangford** [comment]
## Investigation Summary

The non-determinism is caused by **random state not being shared across daemon children**.

### Root Cause
1. When daemon forks children, `weights` and `shared_data` are shared via mmap
2. But `_random_state_sp` (random state) is **NOT shared** - each child gets an independent copy
3. In `sample_pdf.cc`, PDF sampling uses the random state, which diverges across children
4. Different children processing the same input produce different results

### Fix Options
1. **Share random_state via mmap** - Complex, needs atomic operations
2. **Deterministic seed per request** - Hash input to generate deterministic seed (simplest)
3. **Document limitation** - Note that daemon mode with stochastic reductions is non-deterministic

This requires careful architectural consideration and is not a quick fix.
